File: core/src/main/java/pcb/uwu/units/quantity/TimeUnit.java
package pcb.uwu.units.quantity;

import pcb.uwu.core.BaseUnit;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.UnitCounter;

import java.util.function.Function;

public abstract class TimeUnit extends BaseUnit {

	private final UnitCounter unitCounter;

	public TimeUnit(String symbol, String singularName, String pluralName,
					Function<BigDecimalAmount, BigDecimalAmount> translationToCanonical,
					Function<BigDecimalAmount, BigDecimalAmount> translationFromCanonical) {

		super(3, symbol, singularName, pluralName, translationToCanonical, translationFromCanonical);

		this.unitCounter = new UnitCounter(this);
	}

	@Override
	public Class<? extends BaseUnit> getBaseUnitType() {
		return TimeUnit.class;
	}

	@Override
	public UnitCounter getUnitCounter() {
		return unitCounter;
	}
}

File: dimension/src/main/java/pcb/uwu/unit/base/TimeUnit.kt
package pcb.uwu.unit.base

import pcb.uwu.core.Amount
import pcb.uwu.core.BaseUnit
import pcb.uwu.core.UnitCounter
import kotlin.reflect.KClass

open class TimeUnit(symbol: String,
                    singularName: String,
                    pluralName: String,
                    translationToCanonical: (Amount) -> Amount,
                    translationFromCanonical: (Amount) -> Amount)
    : BaseUnit(precedence = 3,
               symbol = symbol,
               singularName = singularName,
               pluralName = pluralName,
               translationToCanonical = translationToCanonical,
               translationFromCanonical = translationFromCanonical)
{
    override val unitCounter
        get() = UnitCounter(this)
    override val baseUnitType: KClass<out BaseUnit> = TimeUnit::class
}
